ext/intl/grapheme/grapheme_string.c

```diff
@@ -918,4 +918,185 @@ PHP_FUNCTION(grapheme_str_split)
+
+	/* When all costs are equal, levenshtein fulfills the requirements of a metric, which means
+	 * that the distance is symmetric. If string1 is shorter than string2 we can save memory (and CPU time)
+	 * by having shorter rows (p1 & p2). */
+	if (ZSTR_LEN(string1) < ZSTR_LEN(string2) && cost_ins == cost_rep && cost_rep == cost_del) {
+		zend_string *tmp = string1;
+		string1 = string2;
+		string2 = tmp;
+	}
+
+	pstr1 = ZSTR_VAL(string1);
+	pstr2 = ZSTR_VAL(string2);
+
+	intl_convert_utf8_to_utf16(&ustring1, &ustring1_len, pstr1, ZSTR_LEN(string1), &ustatus1);
+
+	if (U_FAILURE(ustatus1)) {
+		/* Set global error code. */
+		intl_error_set_code(NULL, ustatus1);
+
+		/* Set error messages. */
+		intl_error_set_custom_msg(NULL, "Error converting input string to UTF-16", 0);
+		if (ustring1) {
+			efree(ustring1);
+		}
+		RETURN_FALSE;
+	}
+
+	intl_convert_utf8_to_utf16(&ustring2, &ustring2_len, pstr2, ZSTR_LEN(string2), &ustatus2);
+
+	if (U_FAILURE(ustatus2)) {
+		/* Set global error code. */
+		intl_error_set_code(NULL, ustatus2);
+
+		/* Set error messages. */
+		intl_error_set_custom_msg(NULL, "Error converting input string to UTF-16", 0);
+		if (ustring2) {
+			efree(ustring2);
+		}
+		if (ustring1) {
+			efree(ustring1);
+		}
+		RETURN_FALSE;
+	}
+
+	UText *ut1 = NULL;
+	UText *ut2 = NULL;
+	UBreakIterator *bi1, *bi2;
+
+	int32_t strlen_1, strlen_2;
+	strlen_1 = grapheme_split_string(ustring1, ustring1_len, NULL, 0);
+	strlen_2 = grapheme_split_string(ustring2, ustring2_len, NULL, 0);
+
+	if (strlen_1 == 0) {
+		efree(ustring1);
+		efree(ustring2);
+		RETURN_LONG(strlen_2 * cost_ins);
+	}
+	if (strlen_2 == 0) {
+		efree(ustring1);
+		efree(ustring2);
+		RETURN_LONG(strlen_1 * cost_del);
+	}
+
+	unsigned char u_break_iterator_buffer1[U_BRK_SAFECLONE_BUFFERSIZE];
+	unsigned char u_break_iterator_buffer2[U_BRK_SAFECLONE_BUFFERSIZE];
+	bi1 = grapheme_get_break_iterator((void*)u_break_iterator_buffer1, &ustatus1);
+	bi2 = grapheme_get_break_iterator((void*)u_break_iterator_buffer2, &ustatus2);
+
+	ut1 = utext_openUTF8(ut1, pstr1, ZSTR_LEN(string1), &ustatus1);
+	ubrk_setUText(bi1, ut1, &ustatus1);
+	ut2 = utext_openUTF8(ut2, pstr2, ZSTR_LEN(string2), &ustatus2);
+	ubrk_setUText(bi2, ut2, &ustatus2);
+
+	p1 = safe_emalloc(strlen_2 + 1, sizeof(zend_long), 0);
+	p2 = safe_emalloc(strlen_2 + 1, sizeof(zend_long), 0);
+
+	for (i2 = 0; i2 <= strlen_2; i2++) {
+		p1[i2] = i2 * cost_ins;
+	}
+
+	int32_t current1 = 0;
+	int32_t current2 = 0;
+	int32_t pos1 = 0;
+	int32_t pos2 = 0;
+	int32_t usrch_pos = 0;
+	while (pos1 != UBRK_DONE) {
+		current1 = ubrk_current(bi1);
+		pos1 = ubrk_next(bi1);
+		if (pos1 == UBRK_DONE) {
+			break;
+		}
+		p2[0] = p1[0] + cost_del;
+		for (i2 = 0, pos2 = 0; pos2 != UBRK_DONE; i2++) {
+			current2 = ubrk_current(bi2);
+			pos2 = ubrk_next(bi2);
+			if (pos2 == UBRK_DONE) {
+				break;
+			}
+			usrch_pos = grapheme_strpos_utf16(pstr1 + current1, pos1 - current1, pstr2 + current2, pos2 - current2, 0, NULL, 0, 0);
```

Comment:
This function call will make the conversion from UTF-8 to UTF-16 _again_ and it seems very inefficient to do it this way. This is also the equivalent of `strrpos` which doesn't seem to make sense here?

Reply:
Yes, indeed.
However, `usearch.h`  requires UTF-16 because `UChar` is UTF-16.
Maybe, possible to use `u_uastrcpy` to `usearch_open`. 
Please give me a time. I'll try it.

Reply:
Unified internal encoding is UTF-16.